features/cellular/framework/mux/cellular_mux.h

```diff
@@ -0,0 +1,561 @@
+/*
+ * Copyright (c) 2006-2017, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MUX_H
+#define MUX_H
+
+#include <stdbool.h>
+#include "FileHandle.h"
+#include "PlatformMutex.h"
+#include "Semaphore.h"
+#include "EventQueue.h"
+
+#define MUX_DLCI_INVALID_ID       0    /* Invalid DLCI ID. Used to invalidate MuxDataService object. */
+#define MUX_CRC_TABLE_LEN         256u /* CRC table length in number of bytes. */
+
+#ifndef MBED_CONF_MUX_DLCI_COUNT
+#define MBED_CONF_MUX_DLCI_COUNT  3u   /* Number of supported DLCI IDs. */
+#endif
+#ifndef MBED_CONF_MUX_BUFFER_SIZE
+#define MBED_CONF_MUX_BUFFER_SIZE 31u  /* Size of TX/Rx buffers in number of bytes. */
+#endif
+
+/* More RAM needs to allocated if more than 4 DLCI ID to be supported see @ref tx_callback_context for details. */
+MBED_STATIC_ASSERT(MBED_CONF_MUX_DLCI_COUNT <= 4u, "");
+
+
+/* @todo:
+I assume that we need to export some kind of #defines for EVENT_SIZE and MAX_EVENT_COUNT (max number of events that can
+be queued at the same time by the module inside EventQueue, so that the application designer can calculate the RAM
+storage requirements correctly at compile time.
+*/
+
+namespace mbed {
+
+class MuxDataService : public FileHandle
+{
+    friend class Mux;
+public:
+
+    /** Enqueue user data for transmission.
+     *
+     *  @note: This is API is only meant to be used for the multiplexer (user) data service tx. Supplied buffer can be
+     *         reused/freed upon call return.
+     *
+     *  @param buffer Begin of the user data.
+     *  @param size   The number of bytes to write.
+     *  @return       The number of bytes written.
+     */
+    virtual ssize_t write(const void* buffer, size_t size);
+
+    /** Read user data into a buffer.
+     *
+     *  @note: This is API is only meant to be used for the multiplexer (user) data service rx.
+     *
+     *  @param buffer The buffer to read in to.
+     *  @param size   The number of bytes to read.
+     *  @return       The number of bytes read, -EAGAIN if no data available for read.
+     */
+    virtual ssize_t read(void *buffer, size_t size);
+
+    /** Not supported by the implementation. */
+    virtual off_t seek(off_t offset, int whence = SEEK_SET);
+
+    /** Not supported by the implementation. */
+    virtual int close();
```

Comment:
Even if this internal file handle doesn't support blocking, it still should support `poll()` so that readable and writable work.

Reply:
Managed by IOTCELL-571